dds/DCPS/security/SSL/SubjectName.h

```diff
@@ -10,17 +10,50 @@
 #include <dds/DCPS/security/OpenDDS_Security_Export.h>
 
 #include <string>
-#include <map>
+#include <vector>
+#include <algorithm>
 
 OPENDDS_BEGIN_VERSIONED_NAMESPACE_DECL
 
 namespace OpenDDS {
 namespace Security {
 namespace SSL {
 
+// A parser for subject names represented by the string format described in RFC 4514.
+class OpenDDS_Security_Export Parser {
+public:
+  typedef std::pair<std::string, std::string> AttributeValueAssertion;
+  typedef std::vector<AttributeValueAssertion> AVAVec;
+
+  Parser(std::string in) : in_(in), pos_(0) {}
+
+  // Parse and store the result to the provided vector.
+  bool parse(AVAVec& store);
+  void reset(std::string in);
+
+private:
+  bool is_alpha(char c) const;
+  bool accept(char c);
+  bool distinguished_name(AVAVec&);
+  bool relative_distinguished_name(AVAVec&);
+  bool attribute_type_value(AVAVec&);
+  bool attribute_type(std::string& at);
+  bool validate_attribute_type(const std::string& at) const;
+
+  bool attribute_value(std::string& av);
+  void unescape(std::string& av) const;
+  void replace_all(std::string& str, const std::string& s, const std::string& t) const;
+  bool validate_attribute_value(const std::string& av) const;
+
+  std::string in_;
+
+  // Current position in the input string.
+  std::string::size_type pos_;
+};
+
 class OpenDDS_Security_Export SubjectName {
- private:
-    typedef std::map<std::string, std::string> AttrMap;
+private:
+  typedef std::vector<std::pair<std::string, std::string> > AttrVec;
```

Comment:
```suggestion
  typedef std::vector<std::map<std::string, std::string> > AttrVec;
```

See PLUS in https://datatracker.ietf.org/doc/html/rfc4514.